runtime/compiler/optimizer/J9TransformUtil.cpp

```diff
@@ -208,6 +209,55 @@ bool J9::TransformUtil::avoidFoldingInstanceField(
       "avoidFoldingInstanceField: symbol %p is never foldable (expected possibly foldable)\n",
       field);
 
+   if (fej9->isStable(field, cpIndex, owningMethod, comp) && !field->isFinal())
+      {
+      uintptr_t fieldAddress = object + fieldOffset;
+
+      TR::DataType loadType = field->getDataType();
+      switch (loadType)
+         {
+         case TR::Int32:
+            {
+            int32_t value = *(int32_t*)fieldAddress;
+            if (value == 0)
+               return true;
+            }
+            break;
+         case TR::Int64:
+            {
+            int64_t value = *(int64_t*)fieldAddress;
+            if (value == 0)
+               return true;
+            }
+            break;
+         case TR::Float:
+            {
+            float value = *(float*)fieldAddress;
+            if (value == 0.0)
+               return true;
+            }
+            break;
+         case TR::Double:
+            {
+            double value = *(double*)fieldAddress;
+            if (value == 0.0)
+               return true;
+            }
+            break;
+         case TR::Address:
+            {
+            TR_ASSERT_FATAL(field->isCollectedReference(), "Expecting a collectable reference\n");
+            uintptr_t value = fej9->getReferenceFieldAtAddress((uintptr_t)fieldAddress);
+            if (value == 0)
+               return true;
+            }
+            break;
+         default:
+            TR_ASSERT_FATAL(false, "Unknown type of field being dereferenced\n");
```

Comment:
I suppose that `field->getDataType()` will give us `TR::Int32` for narrow integral types (`boolean`, `byte`, `short`, `char`)? I know we store those as 32-bit integers - just want to make sure that if one of those is annotated `@Stable` that it doesn't send us into this default case

Reply:
Yeah, I am pretty sure the shadow symbol on those nodes will be TR::Int32.